server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -589,6 +610,112 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
+        final ActionListener<List<StoreFileMetadata>> listener;
+        final CountDown countDown;
+        final BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> pendingSnapshotFilesToRecover;
+        final AtomicBoolean cancelled = new AtomicBoolean();
+        List<StoreFileMetadata> filesFailedToDownloadFromSnapshot;
+
+        SnapshotRecoverFileRequestsSender(ShardRecoveryPlan shardRecoveryPlan, ActionListener<List<StoreFileMetadata>> listener) {
+            this.snapshotFilesToRecover = shardRecoveryPlan.getSnapshotFilesToRecover();
+            this.listener = listener;
+            this.countDown = new CountDown(shardRecoveryPlan.getSnapshotFilesToRecover().size());
+            this.pendingSnapshotFilesToRecover =
+                new LinkedBlockingQueue<>(shardRecoveryPlan.getSnapshotFilesToRecover().getSnapshotFiles());
+        }
+
+        void start() {
+            for (int i = 0; i < maxConcurrentSnapshotFileDownloads; i++) {
+                sendRequest();
+            }
+        }
+
+        void sendRequest() {
+            BlobStoreIndexShardSnapshot.FileInfo snapshotFileToRecover = pendingSnapshotFilesToRecover.poll();
+            if (snapshotFileToRecover == null) {
+                return;
+            }
+
+            try {
+                cancellableThreads.checkForCancel();
+
+                ActionListener<Void> sendRequestListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {
+                        onRequestCompletion(snapshotFileToRecover.metadata(), null);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(new ParameterizedMessage("failed to recover file [{}] from snapshot, " +
+                            "will recover from primary instead", snapshotFileToRecover.metadata()), e);
+                        onRequestCompletion(snapshotFileToRecover.metadata(), e);
+                    }
+                };
+
+                recoveryTarget.downloadSnapshotFile(
+                    snapshotFilesToRecover.getRepository(),
+                    snapshotFilesToRecover.getIndexId(),
+                    snapshotFileToRecover,
+                    ActionListener.runAfter(sendRequestListener, this::sendRequest)
+                );
+            } catch (Exception e) {
+                onFailure(e);
+            }
+        }
+
+        void onFailure(Exception e) {
+            if (cancelled.compareAndSet(false, true)) {
+                pendingSnapshotFilesToRecover.clear();
+                listener.onFailure(e);
+            }
+        }
+
+        void onRequestCompletion(StoreFileMetadata storeFileMetadata, @Nullable Exception exception) {
+            if (cancelled.get()) {
+                return;
+            }
+
+            if (exception != null) {
+                addFileFailedToRecoverFromSnapshot(storeFileMetadata);
```

Comment:
I see advantages both ways. We are retrying (forever?) on source-to-target disconnections and (finitely many times) on target-to-repo disconnections so it's got to be a pretty bad network issue to bubble up here.  Most of the issues I'd expect here will yield fast failures, so IMO we should keep going in case it's transient.

One possible holdup would be if we had to wait for a HTTP client connection from the pool to talk to the repo, IIRC there's a timeout on that wait, and in that case we're probably doing too many recoveries or restores so bailing out would complete the recovery quicker, but I don't think we'd hit this with the default settings so it's kind of self-inflicted too, and we'd see it in the logs anyway.

Reply:
I think the default connection pool for S3 is 50 connections, and we cap the `SNAPSHOT` thread pool to 5 workers, so that might be rare enough?

Reply:
Thanks both, I am OK to keep this as is.